src/program/processor/new_order.rs

```diff
@@ -611,3 +698,70 @@ fn process_multiple_new_orders<'a, 'info>(
 
     Ok(())
 }
+
+fn get_available_balances_for_trader<'a>(
+    market_wrapper: &MarketWrapperMut<'a, Pubkey, FIFOOrderId, FIFORestingOrder, OrderPacket>,
+    trader: &Pubkey,
+    vault_context: Option<&PhoenixVaultContext>,
+    base_lot_size: BaseAtomsPerBaseLot,
+    quote_lot_size: QuoteAtomsPerQuoteLot,
+) -> Result<(BaseLots, QuoteLots), ProgramError> {
+    let (base_lots_free, quote_lots_free) = {
+        let trader_index = market_wrapper
+            .inner
+            .get_trader_index(trader)
+            .ok_or(PhoenixError::TraderNotFound)?;
+        let trader_state = market_wrapper
+            .inner
+            .get_trader_state_from_index(trader_index);
+        (trader_state.base_lots_free, trader_state.quote_lots_free)
+    };
+    let (base_lots_available, quote_lots_available) = match vault_context.as_ref() {
+        None => (base_lots_free, quote_lots_free),
+        Some(ctx) => {
+            let quote_account_atoms = ctx.quote_account.amount().map(QuoteAtoms::new)?;
+            let base_account_atoms = ctx.base_account.amount().map(BaseAtoms::new)?;
+            (
+                base_lots_free + base_account_atoms.unchecked_div(base_lot_size),
+                quote_lots_free + quote_account_atoms.unchecked_div(quote_lot_size),
+            )
+        }
+    };
+    Ok((base_lots_available, quote_lots_available))
+}
+
+fn order_packet_has_sufficient_funds<'a>(
+    market_wrapper: &MarketWrapperMut<'a, Pubkey, FIFOOrderId, FIFORestingOrder, OrderPacket>,
```

Comment:
this and above would be nice if market_wrapper could be marked as not mut where you can still pass in a mut one if you really want to, cause we're not mutating it